File: dotnet/src/AI.Sentinel.AspNetCore/Contract/EvaluationResponseValidator.cs
namespace AI.Sentinel.AspNetCore.Contract;

/// <summary>
/// Client-side validation aligned with Java EvaluationResponseValidator.
/// Failures are contract/transport errors, not risk classifications.
/// </summary>
public static class EvaluationResponseValidator
{
    private static readonly HashSet<string> KnownActions = Enum.GetNames<EnforcementAction>()
        .ToHashSet(StringComparer.Ordinal);

    public static void Validate(EvaluationResponse response, string? expectedCorrelationId)
    {
        ArgumentNullException.ThrowIfNull(response);

        if (response.ContractVersion != EvaluationContractConstants.ContractVersion)
        {
            throw new EvaluationContractException(
                "unsupported response contractVersion: " + response.ContractVersion);
        }

        if (expectedCorrelationId != null
            && !expectedCorrelationId.Equals(response.CorrelationId, StringComparison.Ordinal))
        {
            throw new EvaluationContractException("response correlationId mismatch");
        }

        if (string.IsNullOrWhiteSpace(response.CorrelationId))
        {
            throw new EvaluationContractException("response correlationId is required");
        }

        if (!KnownActions.Contains(response.Action.ToString()))
        {
            throw new EvaluationContractException("unknown response action");
        }

        if (response.AnomalyScore is { } anomaly && !IsFinite(anomaly))
        {
            throw new EvaluationContractException("response anomalyScore must be finite or null");
        }

        if (response.PolicyScore is { } policy && !IsFinite(policy))
        {
            throw new EvaluationContractException("response policyScore must be finite or null");
        }

        var proceedExpected = response.Action is EnforcementAction.ALLOW or EnforcementAction.MONITOR;
        if (response.IsRemoteEvaluationFailure)
        {
            ValidateRemoteFailureShape(response);
        }
        else if (response.Proceed != proceedExpected)
        {
            throw new EvaluationContractException("response proceed inconsistent with action");
        }

        if (response.EvaluationStatuses.Count > 64)
        {
            throw new EvaluationContractException("too many evaluationStatuses");
        }

        foreach (var status in response.EvaluationStatuses)
        {
            if (string.IsNullOrWhiteSpace(status))
            {
                throw new EvaluationContractException("evaluationStatuses entry must be non-blank");
            }

            if (status.Length > EvaluationContractConstants.MaxStringLength)
            {
                throw new EvaluationContractException("evaluationStatuses entry exceeds max length");
            }

            if (!status.Equals(status.ToUpperInvariant(), StringComparison.Ordinal))
            {
                throw new EvaluationContractException("evaluationStatuses must be uppercase codes");
            }
        }

        if (response.Factors.Count > 64)
        {
            throw new EvaluationContractException("too many factors");
        }

        if (response.Advice != null)
        {
            if (string.IsNullOrWhiteSpace(response.Advice.Code))
            {
                throw new EvaluationContractException("advice.code is required");
            }

            var present = response.Factors.Select(f => f.Code).ToHashSet(StringComparer.Ordinal);
            foreach (var linked in response.Advice.LinkedFactorCodes)
            {
                if (!present.Contains(linked))
                {
                    throw new EvaluationContractException(
                        "advice linkedFactorCodes must reference present factors");
                }
            }
        }
    }

    private static void ValidateRemoteFailureShape(EvaluationResponse response)
    {
        if (response.Action != EnforcementAction.ALLOW
            || !response.Proceed
            || response.AnomalyScore != null
            || response.PolicyScore != null
            || response.StartupGraceActive
            || response.Factors.Count > 0
            || response.Advice != null)
        {
            throw new EvaluationContractException(
                "REMOTE_EVALUATION_FAILURE response must be fail-open ALLOW with no scores, factors, or advice");
        }
    }

    private static bool IsFinite(double value) => !double.IsNaN(value) && !double.IsInfinity(value);
}

File: ai-sentinel-core/src/main/java/dev/aisentinel/core/contract/EvaluationResponseValidator.java
package dev.aisentinel.core.contract;

import dev.aisentinel.core.policy.EnforcementAction;

import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Set;

/**
 * Client-side validation of remote {@link EvaluationResponse} payloads.
 * Failures are contract/transport errors, not risk classifications.
 */
public final class EvaluationResponseValidator {

    private static final Set<String> KNOWN_ACTIONS = new HashSet<>();

    static {
        for (EnforcementAction action : EnforcementAction.values()) {
            KNOWN_ACTIONS.add(action.name());
        }
    }

    private EvaluationResponseValidator() {
    }

    public static void validate(EvaluationResponse response, String expectedCorrelationId) {
        Objects.requireNonNull(response, "response");
        if (response.contractVersion() != EvaluationContract.CONTRACT_VERSION) {
            throw new EvaluationContractException(
                "unsupported response contractVersion: " + response.contractVersion());
        }
        if (expectedCorrelationId != null
            && !expectedCorrelationId.equals(response.correlationId())) {
            throw new EvaluationContractException("response correlationId mismatch");
        }
        if (response.correlationId() == null || response.correlationId().isBlank()) {
            throw new EvaluationContractException("response correlationId is required");
        }
        if (response.action() == null) {
            throw new EvaluationContractException("response action is required");
        }
        if (!KNOWN_ACTIONS.contains(response.action().name())) {
            throw new EvaluationContractException("unknown response action");
        }
        if (response.anomalyScore() != null && !Double.isFinite(response.anomalyScore())) {
            throw new EvaluationContractException("response anomalyScore must be finite or null");
        }
        if (response.policyScore() != null && !Double.isFinite(response.policyScore())) {
            throw new EvaluationContractException("response policyScore must be finite or null");
        }
        boolean proceedExpected = switch (response.action()) {
            case ALLOW, MONITOR -> true;
            case THROTTLE, BLOCK, QUARANTINE -> false;
        };
        boolean remoteFailure = response.evaluationStatuses().contains("REMOTE_EVALUATION_FAILURE");
        if (remoteFailure) {
            validateRemoteFailureShape(response);
        } else if (response.proceed() != proceedExpected) {
            throw new EvaluationContractException("response proceed inconsistent with action");
        }
        List<String> statuses = response.evaluationStatuses();
        if (statuses.size() > 64) {
            throw new EvaluationContractException("too many evaluationStatuses");
        }
        for (String status : statuses) {
            if (status == null || status.isBlank()) {
                throw new EvaluationContractException("evaluationStatuses entry must be non-blank");
            }
            if (status.length() > EvaluationContract.MAX_STRING_LENGTH) {
                throw new EvaluationContractException("evaluationStatuses entry exceeds max length");
            }
            if (!status.equals(status.toUpperCase(Locale.ROOT))) {
                throw new EvaluationContractException("evaluationStatuses must be uppercase codes");
            }
        }
        if (response.factors().size() > 64) {
            throw new EvaluationContractException("too many factors");
        }
        if (response.advice() != null) {
            ContractSecurityAdvice advice = response.advice();
            if (advice.code() == null || advice.code().isBlank()) {
                throw new EvaluationContractException("advice.code is required");
            }
            Set<String> present = new HashSet<>();
            for (ContractRiskFactor factor : response.factors()) {
                present.add(factor.code());
            }
            for (String linked : advice.linkedFactorCodes()) {
                if (!present.contains(linked)) {
                    throw new EvaluationContractException("advice linkedFactorCodes must reference present factors");
                }
            }
        }
    }

    private static void validateRemoteFailureShape(EvaluationResponse response) {
        if (response.action() != EnforcementAction.ALLOW || !response.proceed()
            || response.anomalyScore() != null || response.policyScore() != null
            || response.startupGraceActive() || !response.factors().isEmpty()
            || response.advice() != null) {
            throw new EvaluationContractException(
                "REMOTE_EVALUATION_FAILURE response must be fail-open ALLOW with no scores, factors, or advice");
        }
    }
}
